src/HomogeneousCalorimeter_geo.cpp

```diff
@@ -324,18 +376,133 @@ static std::tuple<int, int> add_disk(Detector& desc, Assembly& env, xml::Collect
+  Transform3D tr_global_Iring = RotationZYX(0., 0., 0.) * Translation3D(0., 0., 0.);
+
+  
+  // Supporting frame for the cabling
+  //
+  // PolyhedraRegular cabling_support12(12, r12max, 70.*cm, 25.4/2.*cm);
+  // Box hole(135./2.*mm, 25.4/2.*mm, 25.4/2.*cm);
+  // Rotation3D rot3D(RotationZYX(60.*degree, 0., 0.));
+  
+  // SubtractionSolid cabling_support_subtract_0(cabling_support12, hole, Transform3D(RotationZ(15.*degree) * Position(-82.5*mm, 675.*mm, 0.)));
+  // SubtractionSolid cabling_support_subtract_0(cabling_support12, hole, RotationZYX(-90.*degree, 0., 0.) * Position(-95.013975*mm, 673.352504*mm, 0.));
+  // SubtractionSolid cabling_support_subtract_1(cabling_support_subtract_0, hole, Position(95.013975*mm, -673.352504*mm, 0.));
+  // Volume cabling_support_subtract_V("cabling_support_subtract_V", hole, ring_material);
+  // Transform3D tr_global_cabling_ring = RotationZYX(15.*degree, 0., 0.) * Translation3D(0., 100.*cm, -10.*cm);
+
+  
+  // Place frames and mother volume of modules into the world volume
+  //
+  if (has_envelope) 
+    {
+      env.placeVolume(env_vol, tr_global);
+      env.placeVolume(ring12_vol, tr_global_Oring);
+      env.placeVolume(Sring12_vol, tr_global_Iring);
+
+    }
+
+    
+
+  
+  // local placement of modules
+  //
+  int mid = 0, total_id = 0;  
+  auto points = epic::geo::fillRectangles({0., 0.}, modSize.x(), modSize.y(), rmin, rmax, phimin, phimax);
+
+  Transform3D add_local = RotationZYX(-15.*degree, 0.0, 0.0) * Translation3D(-14.35*cm, 61.5*cm, 0.0);
+  auto modPV = env_vol.placeVolume(modVol, add_local);
+  modPV.addPhysVolID("sector", sector_id).addPhysVolID("module", total_id);
+
+  // Place the modules in the circle
+  for (auto& p : points)
+    {            
+      Transform3D tr_local = RotationZYX(-15.*degree, 0.0, 0.0) * Translation3D(p.x(), p.y(), 0.0);
+      modPV = (has_envelope ? env_vol.placeVolume(modVol, tr_local) : env.placeVolume(modVol, tr_global * tr_local));
+      total_id = id_begin + mid++;
+      modPV.addPhysVolID("sector", sector_id).addPhysVolID("module", total_id);      
+      // std::cout << total_id << ", " << p.x() << ", " << p.y() << std::endl; 
+    }
+
+  // Add the modules manually in the gap
+  const int add_N_mod = 51;  
+  double addX[add_N_mod] = {-16.4,-18.45,-20.5,-38.95,-43.05,-45.1,-47.15,-49.2,-59.45,-61.5,-61.5,-61.5,16.4,18.45,20.5,38.95,43.05,45.1,47.15,49.2,59.45,61.5,61.5,61.5,-16.4,-18.45,-20.5,-38.95,-43.05,-45.1,-47.15,-49.2,-59.45,-61.5,-61.5,-61.5,16.4,18.45,20.5,38.95,43.05,45.1,47.15,49.2,59.45,61.5,61.5,61.5,14.35,14.35,-14.35};
+  double addY[add_N_mod] = {61.5,61.5,59.45,49.2,47.15,45.1,43.05,38.95,20.5,14.35,16.4,18.45,61.5,61.5,59.45,49.2,47.15,45.1,43.05,38.95,20.5,14.35,16.4,18.45,-61.5,-61.5,-59.45,-49.2,-47.15,-45.1,-43.05,-38.95,-20.5,-14.35,-16.4,-18.45,-61.5,-61.5,-59.45,-49.2,-47.15,-45.1,-43.05,-38.95,-20.5,-14.35,-16.4,-18.45,-61.5,61.5,-61.5};
```

Comment:
These are magic positions that need to be determined inside the plugin.

Reply:
Hi, those positions are add by manually to make sure all the modules can be fit into the gap. Should I need to put those numbers to the other scripts?

Reply:
This is probably good enough for now. We try to keep all parameters like these in the xml files instead of compiled code. Changing one of these should not require knowing c++ if it can be avoided.

Reply:
Do you have any example how to write the array in the xml file then declare the array in C++ to read from xml?

For example, I try to do this in xml file:
<constant name="xpos_between_outerframe_gap" value="10., 20."/>

then in the correspondent block I call them with another variable
xpos_add="xpos_between_outerframe_gap"

then in my C++ scripts, I call xpos_add in this way:
double xp_add[2]        = plm.attr<double>(_Unicode(xpos_add));

But I got the error as following:
/homeijclab/pwang/epic/contribution/epic/src/HomogeneousCalorimeter_geo.cpp:417:45: error: array must be initialized with a brace-enclosed initializer
  417 |   double xp_add[2]        = plm.attr<double>(_Unicode(xpos_add));